### ISSUE TITLE
Mermaid diagrams are broken

### ISSUE THREAD

**Issue by NanoSector** [kubernetes/website#31960]
**This is a Bug Report**

**Problem:**
The graphs on the Ingress page are broken in Firefox, but seem to work properly in Chromium. I am unsure if this happens on other pages as well.

Firefox:
![Screenshot_20220228_084718](https://user-images.githubusercontent.com/1280380/155944264-d13ecdce-d54a-41bb-97a3-a4997e49a8e7.png)

Chromium:
![Screenshot_20220228_084802](https://user-images.githubusercontent.com/1280380/155944315-4cc09a95-84f8-4459-b338-a7b8b5560cf2.png)

**Proposed Solution:**
The graphs should be working in Firefox, possibly by fixing the unknown "syntax error".

**Page to Update:**
https://kubernetes.io/docs/concepts/services-networking/ingress/

**Additional Information:**
- Firefox 97.0.1
- Chromium 98
- Ubuntu 20.04.4

**Comment by jihoon-seo** [comment]
For me:
- Ubuntu 18.04, Firefox 97.0 (64-bit)
  (I cannot reproduce on here)
![image](https://user-images.githubusercontent.com/46767780/155951960-1abf65b6-3ce1-408b-88e2-30b62798c929.png)
- Windows 10, 97.0.1 (latest) (64-bit)
  (I cannot reproduce on here)
![image](https://user-images.githubusercontent.com/46767780/155951881-1462ab6d-1bf6-4268-a7e3-96706ac59b4e.png)

I googled and found some links:
- https://addons.mozilla.org/ko/firefox/addon/github-mermaid/
- https://github.com/squidfunk/mkdocs-material/issues/3232
- https://github.com/fralau/mkdocs-mermaid2-plugin/issues/11

**Comment by sftim** [comment]
I can reproduce this using Firefox 97.0 on Ubuntu

/triage accepted
/priority important-longterm
/area web-development

**Comment by NanoSector** [comment]
I can consistently reproduce this on macOS 12.2.1.

<img width="679" alt="image" src="https://user-images.githubusercontent.com/1280380/155955316-a1978afb-2539-4e74-b703-8cf07fcfc1ea.png">

This is a fresh Firefox installation without any add-ons.

However, switching to another network seems to fix the issue. Switching to my phone's tethered 4G network causes it to load properly in Firefox on macOS, switching back to our office network breaks it again.

I cannot find any console logs and network requests seem to successfully go through as well. Downloading mermaid.min.js on both networks and comparing hashes results in the same hash for both files.

Artificially throttling the request to "Regular 4G" using Firefox's developer tools **seems to fix the issue** on our office network. Maybe it is a race condition? [Your first link](https://github.com/squidfunk/mkdocs-material/issues/3232) seems to suggest a fix is published, maybe an update of the site generation tools would fix this?

**Comment by sftim** [comment]
/retitle Mermaid graphs are broken (replicable with Firefox)

**Comment by sftim** [comment]
/retitle Mermaid diagrams are broken (replicable with Firefox)

**Comment by chrismetz09** [comment]
Cannot reproduce using firefox 95.0.2 (64-bit) on mac 10.15.7.
Cannot reproduce using firefox 98.0 on mac 10.15.7.
Cannot reproduce in mermaid live editor that uses Mermaid 8.14.0

@sftim, thoughts on this?

**Comment by sftim** [comment]
I can reproduce this reliably but _not_ on first page load.

**Comment by sftim** [comment]
This is a multihomed PC by the way; 2 IPv4 addresses and 4 IPv6 addresses, so the browser will be using something similar to [Happy Eyeballs](https://en.wikipedia.org/wiki/Happy_Eyeballs).

**Comment by chrismetz09** [comment]
Random number of reloads **DOES** produce the error using firefox 98.0 on mac 10.15.7.
Random number of reloads **DOES** produce the error using safari Version 14.1 (15611.1.21.161.7, 15611) on mac 10.15.7.

I was able to reproduce @NanoSector observation describing firefox 4G throttling that seems to mitigate the problem.

**Comment by chrismetz09** [comment]
@sftim @NanoSector @jihoon-seo 
**Possible fix (but still testing):**
1. [Refreshing page will show Syntax error in mermaid graph sometimes #3232](https://github.com/squidfunk/mkdocs-material/issues/3232) noted `...If Mermaid.js is loaded from cache, Mermaid will sometimes not pick up the configuration and start processing before one can tell it to not startOnLoad. ...`

2. Looked into [Firefox loading pages from cache after latest update](https://support.mozilla.org/en-US/questions/1304951).

3. @sftim observed `I can reproduce this reliably but not on first page load.` Does this mean initial page load IS NOT from cache? I think so? I can reproduce error **after initial page load** by hitting multiple subsequent page reloads. 

4. Tried `shift + Reload` described in [Firefox loading pages from cache after latest update](https://support.mozilla.org/en-US/questions/1304951). I could not reproduce error.

5. There is a [No-Cache No-Store for Selected Hosts add-on](https://addons.mozilla.org/en-US/firefox/addon/no-cache-no-store-for-css/?utm_source=addons.mozilla.org&utm_medium=referral&utm_content=search). Have not tested this yet.

Still testing and will report back.

**Comment by chrismetz09** [comment]
1. Tested [[No-Cache No-Store for Selected Hosts add-on](https://addons.mozilla.org/en-US/firefox/addon/no-cache-no-store-for-css/?utm_source=addons.mozilla.org&utm_medium=referral&utm_content=search).). Did not work. Removed add-on then reproduced error.

Do we want to tell users to install add-on before using?

2. I could not reproduce error using `Shift + command/R` in Firefox.

@sftim, guidance requested. Options I see are:

a) close the issue with a firefox mitigation note. Also add note in [Diagram Guide](https://kubernetes.io/docs/contribute/style/diagram-guide/).

b) Pass it on to web-dev folks? 

tx

**Comment by sftim** [comment]
@chrismetz09 would you be willing to add this to the agenda for the next SIG Docs meeting and / or highlight this via Slack as an issue that needs attention?

**Comment by chrismetz09** [comment]
Will do.

**Comment by reylejano** [comment]
The mermaid diagram is rendering for me, tested on Firefox and Chrome
![image](https://user-images.githubusercontent.com/52760880/159631593-6cff41bb-55b7-4e17-a4cd-db8e4475b8f2.png)

**Comment by chrismetz09** [comment]
Updated to Mermaid 8.14.0 on test branch. 

Replicated previously observed behavior:

1. Chrome renders okay, although you do see rare mis-renders when pounding the reload button.
2. FF shows frequent mis-renders. `Shift + command R` seems to clear the error and correctly renders the diagram.

As mentioined in very rare chrome occurrence and frequent FF reloads, you see the following:

![Screen Shot 2022-04-12 at 10 31 43 AM](https://user-images.githubusercontent.com/5543858/163020968-a2dda6e1-784d-48de-a00a-61a73d1bbc0b.png)

---

Verified checksums to ensure downloaded `mermaid.min.js` file is placed in proper `/website/static/js` directory. 
```
➜  ~ curl https://unpkg.com/mermaid@8.14.0/dist/mermaid.min.js | shasum
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 1080k    0 1080k    0     0  5272k      0 --:--:-- --:--:-- --:--:-- 5298k
bb450af2b99ae0904a7264b06ef9d98017a8f417  -
```
```
➜  ~ shasum mermaid.min.js
bb450af2b99ae0904a7264b06ef9d98017a8f417  mermaid.min.js
```

**Comment by danieltharp** [comment]
Just noting this persists on Firefox 100.0b3 on the first load, and cleared up after a refresh..

**Comment by chrismetz09** [comment]
More investigation steps:

1. Updated to Mermaid 8.14.0 version.
2. Update FF browser to 99.0.1 (64-bit) on Mac Catalina 10.15.7
3. Update Chrome to Version 100.0.4896.88 (Official Build) (x86_64) on Mac Catalina 10.15.7
4. Opened [test PR](https://github.com/kubernetes/website/pull/32940)

Observations: 

1. More frequent FF broken diagrams. Do see very occasional Chrome mermaid breakage but only after rapid/repeated reloads.
2. See same behavior from multiple Mermaid diagrams in [Ingress](https://kubernetes.io/docs/concepts/services-networking/ingress/), [Diagram Guide](https://kubernetes.io/docs/contribute/style/diagram-guide/) and [Using services type nodeport](https://kubernetes.io/docs/tutorials/services/source-ip/#source-ip-for-services-with-type-nodeport)
3. In addition to broken syntax error, I saw new broken diagram renderings.

![Screen Shot 2022-04-14 at 12 23 27 PM](https://user-images.githubusercontent.com/5543858/163466783-b4bb6fe6-4ba6-40ae-b89f-793d63071e43.png)

**Comment by chrismetz09** [comment]
Tested with Mermaid 9.0.0 and observed same behavior.

**Comment by chrismetz09** [comment]
Updates:
1. Tested with clean html file depicting mermaid diagram on local machine. Observed no problems.
2. Need to test [Hugo 0.93.0 diagram support](https://gohugo.io/content-management/diagrams/) but ran into error `failed to render shortcode "api-reference": failed to process shortcode: ... failed: template: shortcodes/api-reference.html:7:17: executing "shortcodes/api-reference.html" at <$page.URL>: can't evaluate field URL in type page.Page`. Will need to hack fix in test branch to complete Hugo 0.93.0 test.
3. Looked over Mermaid code. `errorRenderer.js` hard codes error messages to the error svg:
```
    g.append('text') // text label for the x axis
      .attr('class', 'error-text')
      .attr('x', 1240)
      .attr('y', 250)
      .attr('font-size', '150px')
      .style('text-anchor', 'middle')
      .text('Syntax error in graph');
    g.append('text') // text label for the x axis
      .attr('class', 'error-text')
      .attr('x', 1050)
      .attr('y', 400)
      .attr('font-size', '100px')
      .style('text-anchor', 'middle')
      .text('mermaid version ' + ver);
      ```
      Wondering if solution could be Mermaid PR to add customized error message to Mermaid config.
      4. Would be curious if javascript could force reload when Mermaid diagram is rendered.

**Comment by chrismetz09** [comment]
Used [PR #33073](https://github.com/kubernetes/website/pull/33073) to clear `api-reference shortcode` problem allowing `Hugo 0.93.0` upgrade. Tested with several Mermaid versions (8.11.2, 8.14.0, 9.0.0). Observed same syntax errors with FF (more frequent) and Chrome less frequent.

**Comment by sftim** [comment]
I suspect that some _other_ JS code that we load might interfere with Mermaid. It also might be that a different way of loading Mermaid would hide the symptoms.

Happy to focus on this during KubeCon with anyone who has time to join in.

**Comment by chrismetz09** [comment]
Updates:
1. Worked correctly 100% time when using `disable caching` in dev tools. True for chrome and ff. Presumably this emulates what first-time visitor to page will see?
2. Tried various techniques to reload page from server such as `window.location.reload()`, `history.go(0)`, appending hash to url, etc. Don't see anything out there in JS land that can emulate `empty cache and hard reload`. I don't think manipulating web caches is a strategy for success.

**Comment by sftim** [comment]
I'm minded to try a different technique: drop bits of JavaScript out of the page (by locally editing templates, perhaps) until we find a removal that makes the symptoms disappear. Maybe a similar tactic with simplifying the HTML.

What do folks think, is that a way forward?

**Comment by chrismetz09** [comment]
Okay, I'll give that try. In the meantime, we see popular ingress page occasionally rendering syntax error. To help users resolve this quickly, suggest we add note shortcode to perform hard refresh. Ingress diagram would look like this:

![Screen Shot 2022-05-23 at 11 21 41 AM](https://user-images.githubusercontent.com/5543858/169883358-a2d8efe7-0691-4168-b5b2-499955fffa43.png)

Thoughts?

**Comment by sftim** [comment]
We could add the note _or_ convert that diagram to SVG (and revert it when we have a fix). Either is fine by me.

**Comment by chrismetz09** [comment]
okay I'll go with the note with the note for now.

**Comment by sftim** [comment]
I can no longer reliably replicate these symptoms for the published site. For the deploy preview, the symptoms still show up.

Firefox 100.0.2 and Chromium 102.0.5005.49 on Ubuntu 20.04

**Comment by danieltharp** [comment]
Just as a point of clarification, I didn't need to do a hard reload for it to work, any sort of refresh seemed to make it work.

It does seem reasonable that there's maybe a variable or something being reused by another script causing this sporadic issue.

**Comment by chrismetz09** [comment]
@danieltharp, we see that same behavior at times but not always. I will be converting the frequently visited Ingress diagrams to SVGs until fix can be identified and tested. One possibility is indeed something like a JS variable that is mucking things up. Any ideas on tracking this strongly welcomed.

**Comment by chrismetz09** [comment]
Updates:
1. Installed [docsy-example](https://github.com/google/docsy-example), coded up mermaid diagram (no shortcode) and all worked fine. 
2. updated bootstrap, popper, jquery, mermaid and Hugo (0.93.0 extended) in the appropriate locations in `scripts.html` and `head.html`. Did not work.
3. Tried loading these packages in different order. Did not work.

Don't see syntax error anymore on reloads. But see the following:

![Screen Shot 2022-06-07 at 8 39 01 AM](https://user-images.githubusercontent.com/5543858/172429845-29d151d8-e918-48fb-8a93-4e74756995da.png)

**Comment by sftim** [comment]
> drop bits of JavaScript out of the page (by locally editing templates, perhaps) until we find a removal that makes the symptoms disappear

A related thing to try: start with only Docsy and no other JavaScript, and the CSS (or SCSS) as simple as we can make it, eg plain Docsy. Add Mermaid. See if _that_ works. Assuming it does, gradually add back more of the page<sup>†</sup> until we find what's breaking it.

 † this could be a binary search, based on adding back half of the CSS at a time and seeing which half causes problems for the Mermaid rendering.

**Comment by sftim** [comment]
@chrismetz09 folks from the SIG could maybe plan in some time to pair on this, could that be helpful?

**Comment by chrismetz09** [comment]
Okay, will systematically run thru the iterations beginning with docsy. Should not take long as I have it all setup.

Yes, pairing would be great. Heading over to slack to request assistance. Tx @sftim

**Comment by chrismetz09** [comment]
Also scanned PRs by date range (`is:pr merged:2022-02-01..2022-05-01`) and nothing stood out that would cause problem.

**Comment by chrismetz09** [comment]
Updates:
1. Created [test mermaid k8s repo](https://github.com/chrismetz09/TestMermaidK8s). Starting with bare docsy-example theme and will incrementally build up to k8s website. Objective is to identify what component causes unpredictable Mermaid rendering. 
2. Have not observed any syntax errors in testing.
3. devtools shows a multiple force reflow violations (aka layout thrashing) in console. Might this explain the problem? Looking over performance recordings shows force reflows definitely occurring.

**Comment by sftim** [comment]
PR #34794 looks very promising.

**Comment by dprotaso** [comment]
I just encoutered this today using Firefox 108.0.2 (64-bit) on macOS Monterey v12.6.2

**Comment by sftim** [comment]
/retitle Mermaid diagrams are broken

AIUI, we see this in more than one browser

**Comment by sftim** [comment]
Duplicated by https://github.com/kubernetes/website/issues/39413

**Comment by YoinkBird** [comment]
> Confirmed for me too, using :
> 
>     * chromium `108.0.5359.94-1~deb11u1`
> 
>     * firefox-esr `102.6.0esr-1~deb11u1`
> 
>     * System : debian bullseye
> 
>     * [Sample page](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/)
>       
>       * Chromium
>         ![image](https://user-images.githubusercontent.com/3980933/213699054-cc6369cb-3132-4261-805f-ee6808011b8c.png)
>       * Firefox
>         ![image](https://user-images.githubusercontent.com/3980933/213699245-74b40b67-60c0-47b7-99f0-1e5f2a770812.png)
> 
> 
> As a plus, earlier today I had created like 5 mermaids to document a code. All of them worked fine (on firefox) using an online processor, [like this one](https://mermaid.live).

I encountered this as well; image loads in firefox private window but not a "normal" window.
Disabling the "DuckDuckGo Privacy Essentials" extension allowed the SVG to load.

![image](https://user-images.githubusercontent.com/194444/219809108-1d9ee4d0-ae42-489a-83aa-9268cb8b477d.png)

Not to over-speculate, but as an extra datapoint,  "googletagmanager" is blocked by DuckDuckGo Privacy Essentials in a normal firefox window, but not in the private window. So this _might_ be related to googletagmanager not being able to load:
![image](https://user-images.githubusercontent.com/194444/219809651-712a2907-242e-4a16-a8c8-e7612bd47ba8.png)

**Comment by mrgiles** [comment]
I can confirm that The diagram [How does a HorizontalPodAutoscaler work](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#how-does-a-horizontalpodautoscaler-work) also fails to render in Firefox (112.0.1) in both standard and private modes. 
Note: I don't have the `DuckDuckGo Privacy Essentials` extension installed.

**Comment by sftim** [comment]
If we have duplicates of this issue and either:
- the person who filed the issue is happy to have it closed
- it's been marked as a duplicate for over 15 days

then it's OK to close that duplicate issue - **not** this one - with an explanatory note.

**Comment by sftim** [comment]
Duplicated by #42625

**Comment by junktext** [comment]
FYI: I ran into this bug today and for me I figured out that the Privacy Badger v2023.10.31 extension in Firefox was causing issues with mermaid for some reason. Once I disabled Privacy Badger protection, the diagrams loaded up just fine. Privacy Badger was only detecting 3 potential trackers, which even if I manually allowed those connections to go through, the diagrams would still not work properly.

![Screenshot from 2023-11-22 16-16-21](https://github.com/kubernetes/website/assets/4275824/b07cf4c5-64c2-4d89-bb0d-ec3967bae610)

I was using Firefox v119.0.1 on Ubuntu Linux 22.04. The Kubernetes documentation page I saw this issue on today was here: https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/

Lastly, for what it's worth when the diagram fails to load it mentions that mermaid [v8.13.4](https://github.com/mermaid-js/mermaid/releases/tag/8.13.4) is used, but that was released two years ago on 2021-11-18. As of this writing, mermaid is on v10.6.1 (2023-11-06), so I'm wondering if a version upgrade for mermaid may help fix this bug?

**Comment by sftim** [comment]
Help is very welcome!

**Comment by Gauravpadam** [comment]
I'll help with this, This needs some debugging

**Comment by sftim** [comment]
Looks like we load the Mermaid script in two places. I'm not sure that's helpful.

See https://github.com/kubernetes/website/pull/44644#pullrequestreview-1811623910

**Comment by Gauravpadam** [comment]
Also see https://github.com/kubernetes/website/issues/44768

**Comment by invsblduck** [comment]
Latest report details to help @chrismetz09 debug it:

* Reproducible with latest 64-bit Firefox on Linux, v122.0.
* Reproducible in Private Browsing mode with **NO** extensions enabled.
* Must use non-cached page loads (i.e., bypass browser cache) to eventually get a successful render;
    * Once broken, can never get successful render when reloading from cache.
* No apparent pattern to intermittent flapping;
    * Sometimes it renders every other load;
    * Sometimes it renders every 3rd, 4th, or 5th load;
    * It usually works within < 10 loads ("hard" reloads that bypass cache);
    * It never worked twice in a row.
* Failures never indicated any problem with HTTP requests while inspecting network traffic with FF Web Developer Tools.
* Problem is unrelated to Google Tag Manager (www.googletagmanager.com) because Mermaid will successfully render even when DNS resolution is blackholed for this domain query;
    * I use [Pi-Hole](https://pi-hole.net/) to effectively block this domain.
* I saw errors in the Developer Tools console about:
```
Source map error: Error: request failed with status 404
Resource URL: https://kubernetes.io/js/bootstrap-4.6.1.min.js
Source Map URL: bootstrap.min.js.map
```
&nbsp;&nbsp;&nbsp;&nbsp;...but these errors appeared even when the diagrams rendered successfully, AFAICT.

I have basically ignored Javascript since the 1990s (on purpose), so I am slow at debugging front-end issues across different frameworks and browser platforms.

Hope this helps.

**Comment by dennis-benzinger-hybris** [comment]
Happens to me on Safari 17.4

For example on https://kubernetes.io/docs/concepts/scheduling-eviction/topology-spread-constraints/ :
<img width="1045" alt="image" src="https://github.com/kubernetes/website/assets/6652750/6108255b-2a4b-4d95-ac20-8ead65323a00">

A reload solved it for now

**Comment by sftim** [comment]
Are folks still seeing these symptoms?

**Comment by dennis-benzinger-hybris** [comment]
I wasn't able to reproduce it today

**Comment by allan-kg** [comment]
> Are folks still seeing these symptoms?

I have checked today, and it is not occurring, I tried, both in linux-x86_64 :
- 115.14.0esr
- 97.0 (downloaded from mozilla's ftp)
![image](https://github.com/user-attachments/assets/5b005c16-8a1b-47c0-8377-55db23e7c84c)

The only console log was related to a `button is null` issue.

**Comment by sftim** [comment]
Let's treat this as fixed!

/close
